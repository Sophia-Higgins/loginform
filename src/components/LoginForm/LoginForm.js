import React, { useState } from "react";
import { useHistory } from "react-router-dom";
import { API_BASE_URL, ACCESS_TOKEN_NAME } from "../../constants/apiConstants";
import axios from "axios";

export function LoginForm() {
  let history = useHistory();
  const [state, setState] = useState({
    username: "",
    password: "",
    errorMessage: null,
  });

  const handleChange = (e) => {
    const { id, value } = e.target;
    setState((prevState) => ({
      ...prevState,
      [id]: value,
    }));
  };

  const handleSubmitClick = (e) => {
    e.preventDefault();
    const payload = {
      username: state.username,
      password: state.password,
    };
    axios
      .post(API_BASE_URL + "/login", payload)
      .then((response) => {
        if (response.status === 200) {
          localStorage.setItem(ACCESS_TOKEN_NAME, response.data.token);
          redirectToHome();
        }
      })
      .catch((error) => {
        if (error.response.status === 401) {
          setState((prevState) => ({
            ...prevState,
            errorMessage: "Invalid Username or Password. Please try again.",
          }));
        } else {
          setState((prevState) => ({
            ...prevState,
            errorMessage:
              "Something went wrong...Are you sure you belong in The Good Place?",
          }));
        }
      });
  };

  const redirectToHome = () => {
    history.push({
      pathname: "/home",
      state: {
        username: state.username,
      },
    });
  };

  return (
    <div className="card w-50 login-card mt-5 hv-center">
      <div className="card-header bg-primary text-center text-white">
        Please Login
      </div>
      <form>
        <br />
        <div className="form-group col-lg-12 text-left">
          <label htmlFor="exampleUsername">Username</label>
          <input
            type="username"
            className="form-control"
            id="username"
            placeholder="Username"
            value={state.username}
            onChange={handleChange}
          />
        </div>
        <div className="form-group col-lg-12 text-left">
          <label htmlFor="inputPassword">Password</label>
          <input
            type="password"
            className="form-control"
            id="password"
            placeholder="Password"
            value={state.password}
            onChange={handleChange}
          />
          <br />
          <div
            className="alert alert-danger mt-2"
            style={{ display: state.errorMessage ? "block" : "none" }}
            role="alert"
          >
            {state.errorMessage}
          </div>
          <div className="row justify-content-center">
            <button
              type="submit"
              className="btn btn-primary"
              onClick={handleSubmitClick}
            >
              Login
            </button>
          </div>
        </div>
      </form>
    </div>
  );
}

export default LoginForm;
