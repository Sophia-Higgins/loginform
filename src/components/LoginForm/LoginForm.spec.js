import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import LoginForm from './LoginForm';

let wrapper;

beforeEach(() => {
  wrapper = mount(
    <MemoryRouter>
      <LoginForm />
    </MemoryRouter>
  );
});

describe('<LoginForm />', () => {
  it('should match the snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should have a username field', () => {
    expect(wrapper.find('input[type="username"]').length).toEqual(1);
    expect(wrapper.find('input[type="username"]').prop('value')).toEqual('');
  });

  it('should have a password field', () => {
    expect(wrapper.find('input[type="password"]').length).toEqual(1);
    expect(wrapper.find('input[type="password"]').prop('value')).toEqual('');
  });
});
