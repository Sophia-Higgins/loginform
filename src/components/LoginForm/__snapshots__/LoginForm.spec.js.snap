// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> should match the snapshot 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "kmvlxv",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "kmvlxv",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(LoginForm)>
      <LoginForm
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "kmvlxv",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "kmvlxv",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "key": "kmvlxv",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
      >
        <div
          className="card w-50 login-card mt-5 hv-center"
        >
          <div
            className="card-header bg-primary text-center text-white"
          >
            Please Login
          </div>
          <form>
            <br />
            <div
              className="form-group col-lg-12 text-left"
            >
              <label
                htmlFor="exampleUsername"
              >
                Username
              </label>
              <input
                className="form-control"
                id="username"
                onChange={[Function]}
                placeholder="Username"
                type="username"
                value=""
              />
            </div>
            <div
              className="form-group col-lg-12 text-left"
            >
              <label
                htmlFor="inputPassword"
              >
                Password
              </label>
              <input
                className="form-control"
                id="password"
                onChange={[Function]}
                placeholder="Password"
                type="password"
                value=""
              />
              <br />
              <div
                className="alert alert-danger mt-2"
                role="alert"
                style={
                  Object {
                    "display": "none",
                  }
                }
              />
              <div
                className="row justify-content-center"
              >
                <button
                  className="btn btn-primary"
                  onClick={[Function]}
                  type="submit"
                >
                  Login
                </button>
              </div>
            </div>
          </form>
        </div>
      </LoginForm>
    </withRouter(LoginForm)>
  </Router>
</MemoryRouter>
`;
